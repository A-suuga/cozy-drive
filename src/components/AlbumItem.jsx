import styles from '../styles/albumsList'

import React, { Component } from 'react'
// import classNames from 'classnames'
import { Link, withRouter } from 'react-router'
import { translate } from '../lib/I18n'

import { getPhotoLink } from '../actions/photos'

export class AlbumItem extends Component {
  constructor (props) {
    super(props)

    this.state = {
      isLoading: true
    }

    if (props.album && props.album.photosIds.length) {
      getPhotoLink(props.album.photosIds[0])
        .then(link => this.setState({
          url: link,
          isLoading: false
        }))
    } else {
      this.setState({
        url: '',
        isLoading: false
      })
    }
  }

  render () {
    const { t, album, router } = this.props
    const { isLoading, url } = this.state
    const parentPath = router.location.pathname
    return (
      !isLoading &&
        <div className={styles['pho-album']}>
          <Link
            to={`${parentPath}/${album._id}`}
            className={styles['pho-album-link']}
          >
            <img
              className={styles['pho-album-photo-item']}
              src={url}
            />
            <h2 className={styles['pho-album-title']}>{album.name}</h2>
            <h4 className={styles['pho-album-description']}>
              {t('Albums.album_item_description',
                {smart_count: album.photosIds.length})
              }
            </h4>
          </Link>
        </div>
    )
  }
}

export default translate()(withRouter(AlbumItem))
