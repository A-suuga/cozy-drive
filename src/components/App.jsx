import styles from '../styles/app'

import React from 'react'
import { translate } from '../lib/I18n'
import classNames from 'classnames'

import Sidebar from './Sidebar'

const App = ({ t, children }) => (
  <div className={classNames(styles['pho-wrapper'], styles['coz-sticky'])}>

    <Sidebar />

    <main className={styles['pho-content']}>
      { children }
    </main>
  </div>
)

export default translate()(App)
