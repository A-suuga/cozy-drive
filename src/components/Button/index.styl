@require 'cozy-ui'

@media (max-width: (1023/basefont)rem)
    .dri-btn--more
        display inline-flex

@media (max-width: (768/basefont)rem)
    .dri-btn--more
        border        0
        margin-right  0
        height        em(48px, 14px)
        opacity       .5