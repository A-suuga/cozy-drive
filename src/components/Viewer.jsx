/* global cozy */
import styles from '../styles/viewer'

import React, { Component } from 'react'
import { withRouter } from 'react-router'
import Hammer from 'hammerjs'

import ViewerToolbar from './ViewerToolbar'
import Loading from './Loading'
import ImageLoader from './ImageLoader'

const KEY_CODE_LEFT = 37
const KEY_CODE_RIGHT = 39
const TOOLBAR_HIDE_DELAY = 3000

export class Viewer extends Component {
  constructor (props) {
    super(props)

    this.state = {
      isImageLoading: true,
      ...mapRouteToPhotos(props.photos, props.params),
      hideToolBar: false
    }

    this.navigateToPhoto = this.navigateToPhoto.bind(this)
    this.handleImageLoaded = this.handleImageLoaded.bind(this)

    this.hideToolbarTimeout = null
  }

  componentWillReceiveProps (nextProps) {
    this.setState({...mapRouteToPhotos(nextProps.photos, nextProps.params)})
  }

  componentDidMount () {
    this.onKeyDownCallback = this.onKeyDown.bind(this)
    document.addEventListener('keydown', this.onKeyDownCallback, false)

    this.gesturesHandler = new Hammer(this.viewer)
    this.gesturesHandler.on('swipe', this.onSwipe.bind(this))
    this.gesturesHandler.on('tap', this.toggleToolbar.bind(this))

    this.hideToolbarAfterDelay()
  }

  componentWillUnmount () {
    document.removeEventListener('keydown', this.onKeyDownCallback, false)
    // this.gesturesHandler.destroy()
  }

  onKeyDown (e) {
    if (e.keyCode === KEY_CODE_LEFT) this.navigateToPhoto(this.state.previousID)
    else if (e.keyCode === KEY_CODE_RIGHT) this.navigateToPhoto(this.state.nextID)
  }

  onSwipe (e) {
    if (e.direction === Hammer.DIRECTION_LEFT) this.navigateToPhoto(this.state.nextID)
    else if (e.direction === Hammer.DIRECTION_RIGHT) this.navigateToPhoto(this.state.previousID)
  }

  navigateToPhoto (id) {
    if (this.state.singlePhoto) return

    this.setState({
      isImageLoading: true,
      hideToolBar: true
    })
    const router = this.props.router
    const url = router.location.pathname
    const parentPath = url.substring(0, url.lastIndexOf('/'))
    router.push({ pathname: `${parentPath}/${id}`, query: router.location.query })
  }

  handleImageLoaded () {
    this.setState({ isImageLoading: false })
  }

  toggleToolbar () {
    // invert the current state
    const hidden = !this.state.hideToolBar
    this.setState({ hideToolBar: hidden })
    // if the toolbar is now visible, reset the hide delay
    if (!hidden) this.hideToolbarAfterDelay()
  }

  hideToolbarAfterDelay () {
    clearTimeout(this.hideToolbarTimeout)
    this.hideToolbarTimeout = setTimeout(() => {
      this.setState({ hideToolBar: true })
    }, TOOLBAR_HIDE_DELAY)
  }

  render () {
    const { isImageLoading, previousID, nextID, currentPhoto, singlePhoto, hideToolBar } = this.state
    return (
      <div className={styles['pho-viewer-wrapper']} role='viewer' ref={viewer => { this.viewer = viewer }}>
        <ViewerToolbar hidden={hideToolBar} />
        <div className={styles['pho-viewer-content']}>
          {!singlePhoto && <a role='button' className={styles['pho-viewer-nav-previous']} onClick={() => this.navigateToPhoto(previousID)} />}
          <div className={styles['pho-viewer-photo']}>
            {currentPhoto &&
              <ImageLoader
                photo={currentPhoto}
                onLoad={this.handleImageLoaded}
                src={`${cozy.client._url}${currentPhoto.links.large}`}
              />
            }
            {(!currentPhoto || isImageLoading) &&
              <Loading noMargin color='white' />
            }
          </div>
          {!singlePhoto && <a role='button' className={styles['pho-viewer-nav-next']} onClick={() => this.navigateToPhoto(nextID)} />}
        </div>
      </div>
    )
  }
}

const mapRouteToPhotos = (photos = [], params) => {
  let set = photos.map(photo => photo._id)
  let currentID = params.photoId
  let currentPhotoIndex = set.indexOf(currentID)
  let currentPhoto = photos[currentPhotoIndex]

  let nextID = set[(currentPhotoIndex + 1) % set.length]
  let previousID = set[currentPhotoIndex - 1 >= 0 ? currentPhotoIndex - 1 : set.length - 1]

  return {
    singlePhoto: currentID === previousID && currentID === nextID,
    currentPhoto,
    previousID,
    nextID
  }
}

export default withRouter(Viewer)
