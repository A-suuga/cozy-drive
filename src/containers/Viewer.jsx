import styles from '../styles/viewer'

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import Hammer from 'hammerjs'

import { getPhotoLink } from '../actions/photos'
import { getAlbumPhotos } from '../ducks/albums'
import { getTimelineList } from '../ducks/timeline'

import ViewerToolbar from '../components/ViewerToolbar'
import Loading from '../components/Loading'

const KEY_CODE_LEFT = 37
const KEY_CODE_RIGHT = 39

export class Viewer extends Component {
  constructor (props) {
    super(props)

    this.state = {
      imageUrl: '',
      isLoading: true,
      isImageLoading: true
    }

    this.navigateToPhoto = this.navigateToPhoto.bind(this)
    this.handleImageLoaded = this.handleImageLoaded.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    // get currentPhoto image if currentPhoto defined
    let currentPhoto = this.props.currentPhoto || nextProps.currentPhoto
    if (this.state.imageUrl === '' && currentPhoto !== undefined) {
      getPhotoLink(currentPhoto._id)
        .then(link => {
          this.setState({
            imageUrl: link,
            isLoading: false
          })
        })
    }
  }

  componentDidMount () {
    this.onKeyDownCallback = this.onKeyDown.bind(this)
    document.addEventListener('keydown', this.onKeyDownCallback, false)

    this.gesturesHandler = new Hammer(this.viewer)
    this.gesturesHandler.on('swipe', this.onSwipe.bind(this))
  }

  componentWillUnmount () {
    document.removeEventListener('keydown', this.onKeyDownCallback, false)
    this.gesturesHandler.destroy()
  }

  onKeyDown (e) {
    if (e.keyCode === KEY_CODE_LEFT) this.navigateToPhoto(this.props.previousID)
    else if (e.keyCode === KEY_CODE_RIGHT) this.navigateToPhoto(this.props.nextID)
  }

  onSwipe (e) {
    if (e.direction === Hammer.DIRECTION_LEFT) this.navigateToPhoto(this.props.nextID)
    else if (e.direction === Hammer.DIRECTION_RIGHT) this.navigateToPhoto(this.props.previousID)
  }

  navigateToPhoto (id) {
    this.setState({imageUrl: '', isImageLoading: true, isLoading: true})
    let url = this.props.router.location.pathname
    let parentPath = url.substring(0, url.lastIndexOf('/'))

    this.props.router.push(`${parentPath}/${id}`)
  }

  handleImageLoaded () {
    this.setState({ isImageLoading: false })
  }

  render () {
    const { previousID, nextID, currentPhoto } = this.props
    const { imageUrl, isLoading, isImageLoading } = this.state
    return (
      <div className={styles['pho-viewer-wrapper']} role='viewer' ref={viewer => { this.viewer = viewer }}>
        <ViewerToolbar />
        <div className={styles['pho-viewer-content']}>
          <a role='button' className={styles['pho-viewer-nav-previous']} onClick={() => this.navigateToPhoto(previousID)} />
          <div className={styles['pho-viewer-photo']}>
            {!isLoading &&
              <img
                onLoad={this.handleImageLoaded}
                style={isImageLoading ? 'display:none' : ''}
                alt={currentPhoto.name}
                src={imageUrl}
              />
            }
            {(isLoading || isImageLoading) &&
              <Loading noMargin />
            }
          </div>
          <a role='button' className={styles['pho-viewer-nav-next']} onClick={() => this.navigateToPhoto(nextID)} />
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  let photos = []
  if (ownProps.params.albumId) { // photos from an album
    photos = getAlbumPhotos(state, ownProps.params.albumId).entries
  } else { // all photos (timeline)
    photos = getTimelineList(state).entries
  }
  let set = photos.map(photo => photo._id)
  let currentID = ownProps.params.photoId
  let currentPhotoIndex = set.indexOf(currentID)
  let currentPhoto = photos[currentPhotoIndex]

  let nextID = set[(currentPhotoIndex + 1) % set.length]
  let previousID = set[currentPhotoIndex - 1 >= 0 ? currentPhotoIndex - 1 : set.length - 1]

  return {
    currentPhoto,
    previousID,
    nextID
  }
}

export default connect(
  mapStateToProps
)(withRouter(Viewer))
