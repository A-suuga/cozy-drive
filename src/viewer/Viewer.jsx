import React, { Component } from 'react'

import ViewerControls from './ViewerControls'
import ImageViewer from './ImageViewer'
import AudioViewer from './AudioViewer'
import VideoViewer from './VideoViewer'
import NoViewer from './NoViewer'

import styles from './styles'

const KEY_CODE_LEFT = 37
const KEY_CODE_RIGHT = 39

const isAndroid = () =>
  window.navigator.userAgent &&
  window.navigator.userAgent.indexOf('Android') >= 0
const isIOS = () =>
  window.navigator.userAgent &&
  /iPad|iPhone|iPod/.test(window.navigator.userAgent)
const isMobile = () => isAndroid() || isIOS()

export default class Viewer extends Component {
  componentDidMount() {
    document.addEventListener('keyup', this.onKeyUp, false)
  }

  componentWillUnmount() {
    document.removeEventListener('keyup', this.onKeyUp, false)
  }

  onKeyUp = e => {
    if (e.keyCode === KEY_CODE_LEFT) this.onPrevious()
    else if (e.keyCode === KEY_CODE_RIGHT) this.onNext()
  }

  onNext = () => {
    const { files, currentIndex } = this.props
    if (currentIndex === files.length - 1) {
      return
    }
    const nextIndex = currentIndex + 1
    const nextFile = files[nextIndex]
    this.onChange(nextFile, nextIndex)
  }

  onPrevious = () => {
    const { files, currentIndex } = this.props
    if (currentIndex === 0) {
      return
    }
    const prevIndex = currentIndex - 1
    const prevFile = files[prevIndex]
    this.onChange(prevFile, prevIndex)
  }

  onChange(nextFile, nextIndex) {
    if (this.props.onChange) {
      this.props.onChange(nextFile, nextIndex)
    }
  }

  render() {
    const { files, currentIndex, onClose } = this.props
    const currentFile = files[currentIndex]
    const fileCount = files.length
    const hasPrevious = currentIndex > 0
    const hasNext = currentIndex < fileCount - 1
    return (
      <div className={styles['pho-viewer-wrapper']} role="viewer">
        <ViewerControls
          currentFile={currentFile}
          onClose={onClose}
          hasPrevious={hasPrevious}
          hasNext={hasNext}
          onPrevious={this.onPrevious}
          onNext={this.onNext}
        >
          {this.renderViewer(currentFile)}
        </ViewerControls>
      </div>
    )
  }

  renderViewer(file) {
    if (!file) return null
    const ComponentName = this.getViewerComponentName(file)
    return (
      <ComponentName
        file={file}
        onSwipeLeft={this.onNext}
        onSwipeRight={this.onPrevious}
        onTap={this.showControls}
      />
    )
  }

  getViewerComponentName(file) {
    switch (file.class) {
      case 'image':
        return ImageViewer
      case 'audio':
        return AudioViewer
      case 'video':
        return isMobile() ? NoViewer : VideoViewer
      default:
        return NoViewer
    }
  }
}
