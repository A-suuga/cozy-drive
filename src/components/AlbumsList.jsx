import styles from '../styles/albumsList'

import React from 'react'

import Empty from '../components/Empty'
import AlbumItem from '../components/AlbumItem'

export const AlbumsList = ({ t, f, albums, onServerError }) => {
  if (albums.length === 0) {
    return <Empty emptyType='albums' />
  } else {
    return (
      <div className={styles['pho-album-list']}>
        {albums.map((a) => <AlbumItem album={a} key={a._id} onServerError={onServerError} />)}
      </div>
    )
  }
}

export default AlbumsList
