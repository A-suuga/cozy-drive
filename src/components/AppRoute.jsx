import React from 'react'
import { Route, Redirect } from 'react-router'

import Timeline from '../containers/Timeline'
import AlbumsView from '../containers/AlbumsView'
import AlbumPhotos from '../containers/AlbumPhotos'
import Viewer from '../containers/Viewer'

import App from './App'
export const ComingSoon = () => (<p style='margin-left: 2em'>Coming soon!</p>)

const AppRoute = (
  <Route component={App}>
    <Redirect from='/' to='photos' />
    <Route path='photos' component={Timeline}>
      <Route path=':photoId' component={Viewer} />
    </Route>
    <Route path='albums' component={AlbumsView}>
      <Route path=':albumId' component={AlbumPhotos}>
        <Route path=':photoId' component={Viewer} />
      </Route>
    </Route>
  </Route>
)

export default AppRoute
