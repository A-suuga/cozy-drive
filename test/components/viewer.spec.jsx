'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { mockT } from '../lib/I18n'
import { Viewer } from '../../src/components/Viewer'

const photoObject = {
  _id: '33dda00f0eec15bc3b3c59a615001ac8',
  created_at: '0001-01-01T00:00:00Z',
  name: 'MonImage.jpg',
  size: '150000',
  updated_at: '0001-01-01T00:00:00Z'
}

describe('Viewer component', () => {
  it('should be displayed correctly with the img src computed from photoId', () => {
    const component = shallow(
      <Viewer t={mockT} photoId={photoObject._id} />
    ).node
    expect(component).toMatchSnapshot()
  })
})
