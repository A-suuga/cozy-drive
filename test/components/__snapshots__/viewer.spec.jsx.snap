exports[`Viewer component should be displayed correctly with the img src computed from photoId 1`] = `
<div
  className="pho-viewer-wrapper"
  role="viewer">
  <_translate />
  <div
    className="pho-viewer-content">
    <div
      className="photo-viewer-nav-previous" />
    <div
      className="pho-viewer-photo">
      <img
        src="http://cozy.local:8080/files/download/33dda00f0eec15bc3b3c59a615001ac8" />
    </div>
    <div
      className="photo-viewer-nav-next" />
  </div>
</div>
`;
